src/compiler/parser.ts

```diff
@@ -1811,7 +1811,7 @@ namespace ts {
         function parseEntityName(allowReservedWords: boolean, diagnosticMessage?: DiagnosticMessage): EntityName {
             let entity: EntityName = parseIdentifier(diagnosticMessage);
             while (parseOptional(SyntaxKind.DotToken)) {
-                const node = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);
+                const node: QualifiedName = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);  // !!!

```

Comment:
The exact problem here is the following: We infer the type of `node` from the call to `createNode`. To evaluate that call we need to know the type of `entity` so we can find the type of the `pos` property. To find the type of `entity` we look at the preceding code paths. One leads from the top of the function. In that code path, `entity` has type `Identifier`, which is less than its full declared type (`Identifier | QualifiedName`). Therefore we need to also analyze the second code path that comes from the bottom of the while statement (the loop around case). In that code path, we have the assignment `entity = finishNode(node)` which requires us to know the type of `node`. Boom! Circularity. Not exactly clear what rule we'd introduce to break that circularity.